files/setup_sudoers.sh

```diff
@@ -0,0 +1,55 @@
+#! /bin/bash
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Read the project metadata key named "sudoers" and add each comma separated value to
+# the sudoers file.
+# +TODO - add switch to permit fetching sudoers key from instance metadata
+stdlib::setup_sudoers() {
+  local user user_list sudoers_file
+  user_list="$(stdlib::metadata_get -k 'project/attributes/sudoers')"
+
+  if [[ -z "${user_list}" ]]; then
+    stdlib::debug "Skipping sudoers setup.  The value of the project metadata key named sudoers is empty.  Set sudoers to a comma separated list to enable sudo support, e.g. sudoers=jmccune,pames"
+    return 0
+  fi
+
+  sudoers_file="$(mktemp)"
+  cp /etc/sudoers "${sudoers_file}"
+
+  for user in ${user_list//,/ }; do
+    if grep -q "^${user}"'\b' "${sudoers_file}"; then
+      stdlib::debug "User ${user} is already in /etc/sudoers, taking no action"
+    else
+      stdlib::info "Adding ${user} to /etc/sudoers"
+      echo -e "${user}\tALL= (ALL)\tNOPASSWD: ALL" >>"${sudoers_file}"
+    fi
+  done
+
+  if cmp --silent /etc/sudoers "${sudoers_file}"; then
+    stdlib::debug "No changes necessary to sudoers file, doing nothing."
+    return 0
+  else
+    stdlib::info "Staged changes to sudoers:"
+    diff -U2 /etc/sudoers "${sudoers_file}" || true
```

Comment:
Instead of using `/etc/sudoers` , have you considered using `/etc/sudoers.d/<new sudoers file>` to set sudoers?

Reply:
I've thought about that, but moved forward with what had existed. I think a more robust enhancement might be adding the users to a group and then asserting a line in sudoers that grant permissions to the group. However this might be better enhanced by a follow-on PR. Thoughts?

Reply:
Unless the image is custom, i think that `/etc/sudoers.d/` is preferred way of setting sudoers. I do agree with you that the best thing would be to have a group already created for this.

1. Create `custom_sudoers` group
2. Create `/etc/sudoers.d/20_additional_sudoers` with `custom_sudoers\tALL= (ALL)\tNOPASSWD: ALL`
3. Go though the list of users and verify if they exist. if they exist execute `usermod -aG  custom_sudoers $user`

Reply:
:+1: I like it. Will refactor and commit, tagging you when it's done. thanks!

Reply:
A PR with this behavior should be rejected for the following reasons:

1. Touching groups is outside the scope of this function.  From a support and maintainability perspective, groups get very messy very quickly because of PAM.  Customers frequently have custom PAM configurations, e.g. to support LDAP.  As a result, don't touch groups because doing so makes assumptions which may not hold true.
2. Compatibility with EL6 default behavior should be maintained.

Reply:
On RHEL 6 image the sudo package creates the following files/folders in /etc:
```
# rpm -ql sudo|grep /etc
/etc/pam.d/sudo
/etc/pam.d/sudo-i
/etc/sudo-ldap.conf
/etc/sudo.conf
/etc/sudoers
/etc/sudoers.d```

We can see that `/etc/sudoers.d` is being installed by `sudo.rpm`